### PR TITLE
Fix CloudStackMachine Missing Finalizer

### PR DESCRIPTION
**Pull request by rejoshed**
*Issue #, if available:* N/A

*Description of changes:*
If creating a VM returns an error, then the CSMachine finalizer isn't added to the CR.

When the CAPI machine is deleted, the CSMachine gets garbage collected without deleting a machine.

It's a bit of a strange paradigm that a VM gets created yet the creation method returns an error, but that's the case.

May want to by default just add finalizers in controllers in the future.

*Testing performed:*
Used an `Insane Instance` to incite an error on VM creation.

Watched as CAPC cycled VMs several times without continually creating them.
image
image
image

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: controllers/cloudstackmachine_controller.go [hunk 1]
```diff
@@ -193,8 +193,9 @@ func (r *CloudStackMachineReconciliationRunner) GetOrCreateVMInstance() (retRes
 
 	if err == nil && !controllerutil.ContainsFinalizer(r.ReconciliationSubject, infrav1.MachineFinalizer) { // Fetched or Created?
```
Comment (rejoshed) [inline controllers/cloudstackmachine_controller.go:194]: If there was ever an error in VM creation, the finalizer was then not added.

File: controllers/cloudstackmachine_controller.go [hunk 2]
```diff
@@ -204,7 +205,7 @@ func (r *CloudStackMachineReconciliationRunner) RequeueIfInstanceNotRunning() (r
 		r.Log.Info("Machine instance is Running...")
 		r.ReconciliationSubject.Status.Ready = true
 	} else if r.ReconciliationSubject.Status.InstanceState == "Error" {
-		r.Log.Info("CloudStackMachine VM in error state. Deleting associated Machine.", "csMachine", r.ReconciliationSubject)
+		r.Log.Info("CloudStackMachine VM in error state. Deleting associated Machine.", "csMachine", r.ReconciliationSubject.GetName())
```
Comment (rejoshed) [inline controllers/cloudstackmachine_controller.go:208]: We were logging just too darn much info here. Name is a bit less verbose.

File: controllers/cloudstackmachine_controller.go [hunk 3]
```diff
@@ -259,6 +260,7 @@ func (r *CloudStackMachineReconciliationRunner) ReconcileDelete() (retRes ctrl.R
 		}
 		return ctrl.Result{}, err
 	}
+	r.Log.Info("VM Deleted.")
```
Comment (rejoshed) [inline controllers/cloudstackmachine_controller.go:263]: Helpful to know when a machine was in fact deleted.

File: controllers/utils/utils.go [hunk 4]
```diff
@@ -143,20 +143,20 @@ func GetOwnerOfKind(ctx context.Context, c clientPkg.Client, owned clientPkg.Obj
 		return errors.Errorf(
 			"found more than one GVK for owner when finding owner kind for %s/%s", owned.GetName(), owned.GetNamespace())
 	}
-	kind := gvks[0].Kind
+	gvk := gvks[0]
+
 	for _, ref := range owned.GetOwnerReferences() {
-		if ref.Kind != kind {
+		if ref.Kind != gvk.Kind {
```
Comment (rejoshed) [inline controllers/utils/utils.go:149]: These are just log message fixups that I noticed while troubleshooting.

**Review by maxdrib — APPROVED** [review]
Great catch! How do we start working on unit tests to make sure these edge cases are covered?
